### ISSUE TITLE
 Nextcloud updater did not install the lastest app versions installation

### ISSUE THREAD

**Issue by daniela-waranie** [nextcloud/server#4742]
Recently i migrated from owncloud 8.2.11 to nextcloud 9.0.58 (most up-to-date version, when time of writing), following the steps here: https://help.nextcloud.com/t/solved-migrating-from-owncloud-8-2-1-4-to-nextcloud-9-0-58/12253

Updater updated some apps, too:
- calendar app from 0.8.2 to 1.4.1
- contacts app from 0.5.0.0 to 1.5.1 

But this is not the most up-to-date version of these apps:
calendar = 1.5.2 (https://github.com/nextcloud/calendar/releases)
contacts = 1.5.3 (https://github.com/nextcloud/contacts/releases)

I verified that this server and PHP is able to connect to the internet.

As tflidd suggested in the forum i file an issue now.

Log since migration from owncloud 8.2.11:
```

Debug 	core 	No update found at the Nextcloud appstore for app 166051 	2017-05-08T12:24:00+00:00 	admin
Debug 	core 	No update found at the Nextcloud appstore for app 166055 	2017-05-08T12:24:00+00:00 	admin
Debug 	core 	No update found at the Nextcloud appstore for app 166049 	2017-05-08T12:23:59+00:00 	admin
Error	PHP	Undefined variable: undefinedVariable at /var/www/nextcloud/3rdparty/nikic/php-parser/lib/PhpParser/Lexer.php#70	2017-05-08T09:11:26+00:00	--
Error	PHP	Undefined variable: undefinedVariable at /var/www/nextcloud/3rdparty/nikic/php-parser/lib/PhpParser/Lexer.php#70	2017-05-08T09:11:17+00:00	--
Debug	core	No update found at the Nextcloud appstore for app 166054	2017-05-07T23:37:51+00:00	--
Debug	core	starting upgrade from 8.2.11.1 to 9.0.58.1	2017-05-07T23:37:29+00:00	--
Debug	core	starting upgrade from 8.2.11.1 to 9.0.58.1	2017-05-07T23:33:42+00:00
```

Is 
`Undefined variable: undefinedVariable at /var/www/nextcloud/3rdparty/nikic/php-parser/lib/PhpParser/Lexer.php#70`

related to my problem?

**Comment by daniela-waranie** [comment]
Please have a look here:
https://help.nextcloud.com/t/update-app-update-center-button-is-missing-on-admin-page/12321
Maybe it is related.

**Comment by MorrisJobke** [comment]
Our updater only updates the server and not the apps. For app updates the app management will take care of fetching the latest version from the store.

We plan to do this: #2692 and automatic updates of apps #3182

**Comment by MorrisJobke** [comment]
And you are also running Nextcloud 9. We brought the in our app store in version 10 - right @LukasReschke? In this version the Nextcloud instance will also show you that an update is available. I guess this is what you are looking for.

As a side note: 9.0.58 was the last version of the 9 series and this series is now end of life. So we recommend anyways to update to 10 or 11.

**Comment by MorrisJobke** [comment]
Sorry - another note: https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule this is the page about the releases.

I hope this answered your questions. I will close this here now.

Feel free to ask here again, if something is unclear. We are happy to help and reopen if this turned out to be a bug in one of the latest maintenance versions :)

**Comment by daniela-waranie** [comment]
I'm on a migration path from owncloud 8.2.11 to nextcloud 11.0.3. As recommanded by your documentation i have to migrate from owncloud 8.2.11 to nextcloud 9.0.58 to nextcloud 10.0.5 to nextcloud 11.0.3.

Even if support for nc9 is end-of-life i think "migration-support" should be still available, because nc9 is subject of the recommended migration path.

@MorrisJobke thanks for your fast response. I'm sorry to say that you are wrong. Nextcloud "occ upgrade" is performaing the migration from owncloud 8.2.11 to nextcloud 9.0.58 + it updated the apps. You can see it in the "occ upgrade" console output. I double checked that by replaying the migration on the snapshot of my VM and on a copy of my data dir backup from that point in time.

Updated:
    calendar app from 0.8.2 to 1.4.1
    contacts app from 0.5.0.0 to 1.5.1

When "occ upgrade" updates apps, i expect that it updates each app to the most up-to-date version. In fact it is updating to a less new version. In fact "occ upgrade" was downloading these app versions (because these versions are not part of the zip file: nextcloud-9.0.58.tar.bz2). I guess that "occ upgrade" checks the config/database and then downloads each app that is not bundled within the release zip. If you think it should not perform an update then it should download the same version as it was installed before "occ upgrade" was executed. Else it is performing a app update.

While #2692 is more about to ensure a migration to the next major release is "safe". As that is a very good suggestion / improvement to reduce the pain for all admins. Why should a admin do a "pre-flight" manually when a script can tell him the same within seconds? I do not know the implementation details but i am not sure why there is a "dependency to the server code"... I think "occ upgrade" has a natural dependency to the "config of the current installed server" and to the app store. I guess app store API has methods for "getAppVersionsList" and "isAppVersionCompatibleToServerVersijonY". If the same feature is used in "occ upgrade" and "app page" then it should be a library used by both.

#3182 looks somehow unrelated, because it is about the "app page" - for ongoing maintenance, not on migrating to a major release. Sure these settings could be re-used on "occ upgrade" but that will be a new ticket. 

**In short words:**
"occ upgrade" should 
a) keep the app versions as it is + run a check-availability + warn if app is not available for next major release
b) update the apps to the most up-to-date version if the user agrees check-availability agrees, too

but not as it is doing it today - "half-way app update".

**Comment by daniela-waranie** [comment]
Step 8 in this tutorial:
https://docs.nextcloud.com/server/10/admin_manual/maintenance/manual_upgrade.html
says:

> If you are using 3rd party applications, look in your new nextcloud/apps/ directory to see if they are there. If not, copy them from your old apps/ directory to your new one. Make sure the directory permissions of your third party application directories are the same as for the other ones.

Figuring out what app is an 3rd party app and with is not is an errorprone process. I my case i failed to copy contacts app and calandar app. This makes "occ upgrade" think the apps should be downloaded - but "occ upgrade" fails to download the most up-to-date or exact the same version.

"occ upgrade" should warn and stop if it detects that a app in filesystem is missing but referenced by config. Or resolve that issue as described above.

**Comment by skjnldsv** [comment]
As I cannot reproduce the original issue anymore, I will close this ticket. If this is still happening please make sure to **upgrade** to the **latest** version. After that, feel free to reopen.
